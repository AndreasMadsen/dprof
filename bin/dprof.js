#!/bin/env node

var endpoint = require('endpoint');
var extend = require('util-extend');

process.stdin.pipe(endpoint(function (err, dprof) {
  if (err) throw err;
  dprof = JSON.parse(dprof);

  var content = new Node(dprof.root, dprof.total, 1, new State(), new Root());

  var svg = new Element('svg').attr({ xmlns: 'http://www.w3.org/2000/svg', width: 1000 });

  process.stdout.write(svg.start() + '\n');
  process.stdout.write(content.stringify() + '\n');
  process.stdout.write(svg.end());
}));

function Root() {
  this.top = 0;
}

function State() {
  this.y = 0;
}

function Node(node, total, depth, state, parent) {
  this.name = node.name;
  this.init = Math.floor(node.init / total * 1000);
  this.before = Math.floor((node.init + node.before) / total * 1000);
  this.after = Math.floor((node.init + node.after) / total * 1000);
  this.stack = node.stack;

  this.depth = depth;
  this.state = state;
  this.parent = parent;
  this.top = state.y;

  this.elems = [];
  this.elems.push(new Element('rect').attr({
    x: this.init,
    y: this.top,
    width: this.before - this.init,
    height: 20,
    fill: 'SteelBlue',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('rect').attr({
    x: this.before,
    y: this.top,
    width: this.after - this.before,
    height: 20,
    fill: 'IndianRed',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('line').attr({
    x1: this.init,
    y1: this.parent.top + 20,
    x2: this.init,
    y2: this.top + 20,
    style: 'stroke: #000; stroke-width: 2px'
  }));

  this.state.y += 20;
  this.children = node.children.map(function (child) {
    return new Node(child, total, depth + 1, state, this);
  }, this);
}

Node.prototype._indent = function () {
  var s = '';
  for (var i = 0; i < this.depth; i++) {
    s += '  ';
  }
  return s;
};

Node.prototype.stringify = function () {
  var content = [];
  this.elems.forEach(function (elem) {
    content.push(this._indent() + elem.singleton());
  }, this);
  this.children.forEach(function (child) {
    content.push(child.stringify());
  }, this);

  return content.join('\n');
};

function Element(name) {
  this._name = name;
  this._attr = {};
}

Element.prototype.attr = function (attr) {
  extend(this._attr, attr);
  return this;
};

Element.prototype.start = function () {
  var start = '<' + this._name;
  Object.keys(this._attr).forEach(function (k) {
    start += ' ' + k + '="' + this._attr[k] + '"';
  }, this);
  start += '>';
  return start;
};

Element.prototype.end = function () {
  return '</' + this._name + '>';
};

Element.prototype.singleton = function () {
  return this.start() + this.end();
};
