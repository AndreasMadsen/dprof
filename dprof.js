'use strict';

const providers = process.binding('async_wrap').Providers;
const asyncWrap = require('./async_wrap.js');
const zlib = require('zlib');
const fs = require('fs');

const version = require('./package.json').version;
const processStart = process.hrtime();

// Change the default stackTraceLimit if it haven't allready been overwritten
if (process.execArgv.indexOf('--stack_trace_limit') === -1 && Error.stackTraceLimit === 10) {
  Error.stackTraceLimit = 8;
}

//
// Define node class
//

function Site(site) {
  this.description = site.toString();
  this.filename = site.getFileName();
  this.line = site.getLineNumber();
  this.collum = site.getColumnNumber();
}

function timestamp() {
  const t = process.hrtime(processStart);
  return t[0] * 1e9 + t[1];
}

function Node(name, stack) {
  this.name = name;
  this._init = timestamp();
  this._before = [];
  this._after = [];
  this.children = [];
  this.stack = stack.map(function (site) {
    return new Site(site);
  });
}

Node.prototype.add = function (handle) {
  const node = new Node(handle.constructor.name, asyncWrap.stackTrace(3));
  this.children.push(node);
  return node;
};

Node.prototype.before = function () {
  this._before.push(timestamp());
};

Node.prototype.after = function () {
  this._after.push(timestamp());
};

Node.prototype.toJSON = function () {
  return {
    name: this.name,
    init: this._init,
    before: this._before,
    after: this._after,
    stack: this.stack,
    children: this.children
  };
};

Node.prototype.rootIntialize = function () {
  this._init = 0;
  this._before.push(0);
};

//
// Setup hooks
//

asyncWrap.setup(asyncInit, asyncBefore, asyncAfter);

const root = new Node('root', asyncWrap.stackTrace(2));
      root.rootIntialize();
let state = root;

function asyncInit(provider, parent) {
  if (provider === providers.TIMERWRAP) {
    this._dprofIgnore = true;
    return;
  }

  if (parent) {
    this._dprofState = parent._dprofState.add(this, provider, parent);
  } else {
    this._dprofState = state.add(this, provider, parent);
  }
}

function asyncBefore() {
  if (this._dprofIgnore) return;

  this._dprofState.before();
  state = this._dprofState;
}

function asyncAfter() {
  if (this._dprofIgnore) return;
  
  this._dprofState.after();
  state = root;
}

// The root job is done when process.nextTick is called
asyncWrap.disable();
process.nextTick(function () {
  root.after();
});
asyncWrap.enable();

//
// Print result
//

process.on('exit', function () {
  // even though zlib is sync, it still fires async_wrap events,
  // so disable asyncWrap just to be sure.
  asyncWrap.disable();

  const data = {
    'total': timestamp(),
    'version': version,
    'root': root
  };

  if (process.env.NODE_DPROF_DEBUG) {
    fs.writeFileSync('./dprof.json', JSON.stringify(data, null, 1));
  } else {
    fs.writeFileSync('./dprof.json.gz', zlib.gzipSync(JSON.stringify(data)));
  }
});
