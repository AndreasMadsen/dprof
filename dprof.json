{
 "name": "root",
 "init": 0,
 "before": 0,
 "after": 4319266,
 "children": [
  {
   "name": "Timer",
   "init": 2044586,
   "before": 548667,
   "after": 1409916,
   "children": [
    {
     "name": "FSReqWrap",
     "init": 3039113,
     "before": 458225,
     "after": 788615,
     "children": [
      {
       "name": "FSReqWrap",
       "init": 3776193,
       "before": 164991,
       "after": 213630,
       "children": []
      },
      {
       "name": "FSReqWrap",
       "init": 3803875,
       "before": 188243,
       "after": 268189,
       "children": [
        {
         "name": "FSReqWrap",
         "init": 4063866,
         "before": 39850,
         "after": 74510,
         "children": []
        }
       ]
      }
     ]
    },
    {
     "name": "FSReqWrap",
     "init": 3279865,
     "before": 623396,
     "after": 650968,
     "children": []
    }
   ]
  }
 ]
}