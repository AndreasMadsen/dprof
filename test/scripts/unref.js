'use strict';
require('../../dprof.js');

const net = require('net');

const server = net.createServer(function (socket) {
  server.close();
  
  socket._handle.unref();
  setTimeout(function () {
    socket._handle.ref();
  }, 100);
  setTimeout(function () {
    socket._handle.unref();
  }, 200);
  setTimeout(function () {
    // keep alive
  }, 300);
});

server.listen(0, 'localhost', function () {
  const addr = server.address();
  const socket = net.connect(addr.port, addr.address, function () {
    socket._handle.unref();
  });
});
