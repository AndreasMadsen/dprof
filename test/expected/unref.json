{
 "stdout": "",
 "stderr": "",
 "dump": {
  "total": 335948057,
  "version": "0.16.0",
  "root": {
   "name": "root",
   "parent": null,
   "uid": 0,
   "init": 0,
   "destroy": 12733556,
   "before": [
    0
   ],
   "after": [
    12682655
   ],
   "unref": [],
   "ref": [],
   "initRef": true,
   "stack": [
    {
     "description": "Module._compile (module.js:541:32)",
     "filename": "module.js",
     "line": 541,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:550:10)",
     "filename": "module.js",
     "line": 550,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:458:32)",
     "filename": "module.js",
     "line": 458,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:417:12)",
     "filename": "module.js",
     "line": 417,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:409:3)",
     "filename": "module.js",
     "line": 409,
     "collum": 3
    },
    {
     "description": "Module.require (module.js:468:17)",
     "filename": "module.js",
     "line": 468,
     "collum": 17
    },
    {
     "description": "require (internal/module.js:20:19)",
     "filename": "internal/module.js",
     "line": 20,
     "collum": 19
    },
    {
     "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:2:1)",
     "filename": "/test/scripts/unref.js",
     "line": 2,
     "collum": 1
    },
    {
     "description": "Module._compile (module.js:541:32)",
     "filename": "module.js",
     "line": 541,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:550:10)",
     "filename": "module.js",
     "line": 550,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:458:32)",
     "filename": "module.js",
     "line": 458,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:417:12)",
     "filename": "module.js",
     "line": 417,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:409:3)",
     "filename": "module.js",
     "line": 409,
     "collum": 3
    },
    {
     "description": "Module.runMain (module.js:575:10)",
     "filename": "module.js",
     "line": 575,
     "collum": 10
    },
    {
     "description": "run (bootstrap_node.js:352:7)",
     "filename": "bootstrap_node.js",
     "line": 352,
     "collum": 7
    },
    {
     "description": "startup (bootstrap_node.js:144:9)",
     "filename": "bootstrap_node.js",
     "line": 144,
     "collum": 9
    },
    {
     "description": "bootstrap_node.js:467:3",
     "filename": "bootstrap_node.js",
     "line": 467,
     "collum": 3
    }
   ],
   "children": [
    1,
    2
   ]
  },
  "nodes": [
   {
    "name": "Signal",
    "parent": 0,
    "uid": 1,
    "init": 3010981,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [
     3703284
    ],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "process.<anonymous> (internal/process.js:195:18)",
      "filename": "internal/process.js",
      "line": 195,
      "collum": 18
     },
     {
      "description": "emitTwo (events.js:106:13)",
      "filename": "events.js",
      "line": 106,
      "collum": 13
     },
     {
      "description": "process.emit (events.js:191:7)",
      "filename": "events.js",
      "line": 191,
      "collum": 7
     },
     {
      "description": "_addListener (events.js:226:14)",
      "filename": "events.js",
      "line": 226,
      "collum": 14
     },
     {
      "description": "process.addListener (events.js:270:10)",
      "filename": "events.js",
      "line": 270,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/dprof.js:184:11)",
      "filename": "/dprof.js",
      "line": 184,
      "collum": 11
     },
     {
      "description": "Module._compile (module.js:541:32)",
      "filename": "module.js",
      "line": 541,
      "collum": 32
     },
     {
      "description": "Object.Module._extensions..js (module.js:550:10)",
      "filename": "module.js",
      "line": 550,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:458:32)",
      "filename": "module.js",
      "line": 458,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:417:12)",
      "filename": "module.js",
      "line": 417,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:409:3)",
      "filename": "module.js",
      "line": 409,
      "collum": 3
     },
     {
      "description": "Module.require (module.js:468:17)",
      "filename": "module.js",
      "line": 468,
      "collum": 17
     },
     {
      "description": "require (internal/module.js:20:19)",
      "filename": "internal/module.js",
      "line": 20,
      "collum": 19
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:2:1)",
      "filename": "/test/scripts/unref.js",
      "line": 2,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:541:32)",
      "filename": "module.js",
      "line": 541,
      "collum": 32
     },
     {
      "description": "Object.Module._extensions..js (module.js:550:10)",
      "filename": "module.js",
      "line": 550,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:458:32)",
      "filename": "module.js",
      "line": 458,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:417:12)",
      "filename": "module.js",
      "line": 417,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:409:3)",
      "filename": "module.js",
      "line": 409,
      "collum": 3
     },
     {
      "description": "Module.runMain (module.js:575:10)",
      "filename": "module.js",
      "line": 575,
      "collum": 10
     },
     {
      "description": "run (bootstrap_node.js:352:7)",
      "filename": "bootstrap_node.js",
      "line": 352,
      "collum": 7
     },
     {
      "description": "startup (bootstrap_node.js:144:9)",
      "filename": "bootstrap_node.js",
      "line": 144,
      "collum": 9
     },
     {
      "description": "bootstrap_node.js:467:3",
      "filename": "bootstrap_node.js",
      "line": 467,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "GetAddrInfoReqWrap",
    "parent": 0,
    "uid": 2,
    "init": 12034877,
    "destroy": 26707888,
    "before": [
     15097699
    ],
    "after": [
     22243027
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "Object.lookup (dns.js:167:19)",
      "filename": "dns.js",
      "line": 167,
      "collum": 19
     },
     {
      "description": "listenAfterLookup (net.js:1393:20)",
      "filename": "net.js",
      "line": 1393,
      "collum": 20
     },
     {
      "description": "Server.listen (net.js:1389:5)",
      "filename": "net.js",
      "line": 1389,
      "collum": 5
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:21:8)",
      "filename": "/test/scripts/unref.js",
      "line": 21,
      "collum": 8
     },
     {
      "description": "Module._compile (module.js:541:32)",
      "filename": "module.js",
      "line": 541,
      "collum": 32
     },
     {
      "description": "Object.Module._extensions..js (module.js:550:10)",
      "filename": "module.js",
      "line": 550,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:458:32)",
      "filename": "module.js",
      "line": 458,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:417:12)",
      "filename": "module.js",
      "line": 417,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:409:3)",
      "filename": "module.js",
      "line": 409,
      "collum": 3
     },
     {
      "description": "Module.runMain (module.js:575:10)",
      "filename": "module.js",
      "line": 575,
      "collum": 10
     },
     {
      "description": "run (bootstrap_node.js:352:7)",
      "filename": "bootstrap_node.js",
      "line": 352,
      "collum": 7
     },
     {
      "description": "startup (bootstrap_node.js:144:9)",
      "filename": "bootstrap_node.js",
      "line": 144,
      "collum": 9
     },
     {
      "description": "bootstrap_node.js:467:3",
      "filename": "bootstrap_node.js",
      "line": 467,
      "collum": 3
     }
    ],
    "children": [
     3,
     -1
    ]
   },
   {
    "name": "TCP",
    "parent": 2,
    "uid": 3,
    "init": 20622808,
    "destroy": 30134720,
    "before": [
     27393728
    ],
    "after": [
     30128485
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "createServerHandle (net.js:1179:14)",
      "filename": "net.js",
      "line": 1179,
      "collum": 14
     },
     {
      "description": "Server._listen2 (net.js:1236:14)",
      "filename": "net.js",
      "line": 1236,
      "collum": 14
     },
     {
      "description": "listen (net.js:1288:10)",
      "filename": "net.js",
      "line": 1288,
      "collum": 10
     },
     {
      "description": "net.js:1398:9",
      "filename": "net.js",
      "line": 1398,
      "collum": 9
     },
     {
      "description": "GetAddrInfoReqWrap.asyncCallback [as callback] (dns.js:65:16)",
      "filename": "dns.js",
      "line": 65,
      "collum": 16
     },
     {
      "description": "GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:84:10)",
      "filename": "dns.js",
      "line": 84,
      "collum": 10
     }
    ],
    "children": [
     6,
     -3,
     -4,
     -5
    ]
   },
   {
    "name": "NextTickWrap",
    "parent": 2,
    "uid": -1,
    "init": 21953830,
    "destroy": 26223908,
    "before": [
     22350756
    ],
    "after": [
     26126277
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "Server._listen2 (net.js:1266:11)",
      "filename": "net.js",
      "line": 1266,
      "collum": 11
     },
     {
      "description": "listen (net.js:1288:10)",
      "filename": "net.js",
      "line": 1288,
      "collum": 10
     },
     {
      "description": "net.js:1398:9",
      "filename": "net.js",
      "line": 1398,
      "collum": 9
     },
     {
      "description": "GetAddrInfoReqWrap.asyncCallback [as callback] (dns.js:65:16)",
      "filename": "dns.js",
      "line": 65,
      "collum": 16
     },
     {
      "description": "GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:84:10)",
      "filename": "dns.js",
      "line": 84,
      "collum": 10
     }
    ],
    "children": [
     4,
     -2
    ]
   },
   {
    "name": "TCP",
    "parent": -1,
    "uid": 4,
    "init": 24573788,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [
     26904164
    ],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "Socket.connect (net.js:913:40)",
      "filename": "net.js",
      "line": 913,
      "collum": 40
     },
     {
      "description": "Object.exports.connect.exports.createConnection (net.js:70:35)",
      "filename": "net.js",
      "line": 70,
      "collum": 35
     },
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:23:22)",
      "filename": "/test/scripts/unref.js",
      "line": 23,
      "collum": 22
     },
     {
      "description": "Server.g (events.js:286:16)",
      "filename": "events.js",
      "line": 286,
      "collum": 16
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "emitListeningNT (net.js:1278:10)",
      "filename": "net.js",
      "line": 1278,
      "collum": 10
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:71:11)",
      "filename": "internal/process/next_tick.js",
      "line": 71,
      "collum": 11
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": []
   },
   {
    "name": "NextTickWrap",
    "parent": -1,
    "uid": -2,
    "init": 26017549,
    "destroy": 26698184,
    "before": [
     26286023
    ],
    "after": [
     26680574
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "lookupAndConnect (net.js:961:13)",
      "filename": "net.js",
      "line": 961,
      "collum": 13
     },
     {
      "description": "Socket.connect (net.js:929:5)",
      "filename": "net.js",
      "line": 929,
      "collum": 5
     },
     {
      "description": "Object.exports.connect.exports.createConnection (net.js:70:35)",
      "filename": "net.js",
      "line": 70,
      "collum": 35
     },
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:23:22)",
      "filename": "/test/scripts/unref.js",
      "line": 23,
      "collum": 22
     },
     {
      "description": "Server.g (events.js:286:16)",
      "filename": "events.js",
      "line": 286,
      "collum": 16
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "emitListeningNT (net.js:1278:10)",
      "filename": "net.js",
      "line": 1278,
      "collum": 10
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:71:11)",
      "filename": "internal/process/next_tick.js",
      "line": 71,
      "collum": 11
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": [
     5
    ]
   },
   {
    "name": "TCPConnectWrap",
    "parent": -2,
    "uid": 5,
    "init": 26545993,
    "destroy": 27290180,
    "before": [
     26734127
    ],
    "after": [
     27282740
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "connect (net.js:854:26)",
      "filename": "net.js",
      "line": 854,
      "collum": 26
     },
     {
      "description": "net.js:963:9",
      "filename": "net.js",
      "line": 963,
      "collum": 9
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:67:7)",
      "filename": "internal/process/next_tick.js",
      "line": 67,
      "collum": 7
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": []
   },
   {
    "name": "TCP",
    "parent": 3,
    "uid": 6,
    "init": 27367249,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [
     28729250,
     232043725
    ],
    "ref": [
     132957304
    ],
    "initRef": true,
    "stack": [],
    "children": []
   },
   {
    "name": "TimeoutWrap",
    "parent": 3,
    "uid": -3,
    "init": 28926364,
    "destroy": 133165938,
    "before": [
     132838320
    ],
    "after": [
     132974427
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:10:3)",
      "filename": "/test/scripts/unref.js",
      "line": 10,
      "collum": 3
     },
     {
      "description": "emitOne (events.js:96:13)",
      "filename": "events.js",
      "line": 96,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:188:7)",
      "filename": "events.js",
      "line": 188,
      "collum": 7
     },
     {
      "description": "TCP.onconnection (net.js:1458:8)",
      "filename": "net.js",
      "line": 1458,
      "collum": 8
     }
    ],
    "children": []
   },
   {
    "name": "TimeoutWrap",
    "parent": 3,
    "uid": -4,
    "init": 29699487,
    "destroy": 232074610,
    "before": [
     231949754
    ],
    "after": [
     232055028
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:13:3)",
      "filename": "/test/scripts/unref.js",
      "line": 13,
      "collum": 3
     },
     {
      "description": "emitOne (events.js:96:13)",
      "filename": "events.js",
      "line": 96,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:188:7)",
      "filename": "events.js",
      "line": 188,
      "collum": 7
     },
     {
      "description": "TCP.onconnection (net.js:1458:8)",
      "filename": "net.js",
      "line": 1458,
      "collum": 8
     }
    ],
    "children": []
   },
   {
    "name": "TimeoutWrap",
    "parent": 3,
    "uid": -5,
    "init": 30069882,
    "destroy": 335639539,
    "before": [
     335575190
    ],
    "after": [
     335627863
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/unref.js:16:3)",
      "filename": "/test/scripts/unref.js",
      "line": 16,
      "collum": 3
     },
     {
      "description": "emitOne (events.js:96:13)",
      "filename": "events.js",
      "line": 96,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:188:7)",
      "filename": "events.js",
      "line": 188,
      "collum": 7
     },
     {
      "description": "TCP.onconnection (net.js:1458:8)",
      "filename": "net.js",
      "line": 1458,
      "collum": 8
     }
    ],
    "children": []
   }
  ]
 }
}
